src/Tasks.UnitTests/Copy_Tests.cs

```diff
@@ -112,6 +121,353 @@ public void Dispose()
             Copy.RefreshInternalEnvironmentValues();
         }
 
+        [Fact]
+        public void CopyWithNoInput()
+        {
+            var task = new Copy { BuildEngine = new MockEngine(true), };
+            task.Execute().ShouldBeTrue();
```

Comment:
Shouldn't the task return false when there are no source folders or files? I am confused here.

Reply:
The existing code returns true when there are no `SourceFiles`.

https://github.com/dotnet/msbuild/blob/ed404c0d68dc073a09c5c9402a5c56f65e7a29a8/src/Tasks/Copy.cs#L400-L411

However, this test is never true when the task is used in MSBuild code because the `SourceFiles` parameter is `[Required]`. Because either or both `SourceFiles` and `SourceFolders` can be provided, neither can be `[Required]`.

To be more consistent with the existing behavior of the `Copy` task, I'll make the "empty set" an error.

Reply:
changes have been committed

Reply:
`[Required]` doesn't distinguish between a null list and an empty list and doesn't allow an empty list. Failing on an empty list actually matches the current behavior of the task when used from MSBuild.

The behavior change is only when the task is used directly in code bypassing the `[Requireed]` logic.

Reply:
> Sorry for my long absence, @jrdodds, I was sick. ...

Sorry to hear you were sick and I hope you have fully recovered.

Reply:
Restored the current behavior. An empty set for sources is accepted and the task returns `true`.